### ISSUE TITLE
The source code is availabe?

### ISSUE THREAD

**Issue by sl1pkn07** [crmarsh/discord-linux-bugs#168]
Hi

the source code is available?

i want build myself for avoid library version missmatch and use system libraries instead of diplicate libs in my system

greetings

**Comment by kessanro** [comment]
@sl1pkn07 i know bit about why he is not posting the source
A: security it would be dangerous to post source code with encryption info because HACKERS and people getting into accounts i learned this the hard way.
B: it might cause people to copy and clone discord and add ads viruses etc.
C: would just be a bad idea because of what i talked about

**Comment by gonzalezjo** [comment]
@kessanro I respectfully disagree. Understand that the Discord client is written in JavaScript, an interpreted language. It's not compiled to JavaScript via something like Emscripten that produces hard to read code, it's just plain old JavaScript. What this means is that it's pretty easy to read the source code. This isn't a bad thing at all though. There's a community of people who have read the JavaScript to try and make Discord more flexible. Check out [BetterDiscord](https://betterdiscord.net/home/) to see what I mean. 


I don't think anyone has a problem with this. However, you express concern that this might make things easier for hackers. You're not wrong, but that's not exactly a bad thing. The idea of keeping things hidden for the purpose of security is known as "Security through obscurity." This isn't always a bad thing and in the case of software like Denuvo, has been wildly successful. But for interpreted languages, this doesn't just not apply, it also doesn't really help. 


A quote I really like explains why: "Rogues are very keen in their profession, and know already much more than we can teach them." There are a lot of people out there trying to help patch and fix bugs. Hackers on the other hand, spend their life dealing with security through obscurity and other techniques meant to stop them, and do so successfully. This scares away the good guys but it's business as usual for the bad guys. 



Discord, as well as just about all of the web (including GitHub!) uses asymmetric cryptography. It's a really complicated concept, but it means that websites are able to share these encryption keys with anyone who visits (that's what the green lock in your URL bar means!) that you can't actually use to decrypt messages, only encrypt. The worlds smartest cryptographers back these algorithms and so far they seem mathematically sound. This means the only way for a hacker to get value out of Discord's encryption algorithms are to hack the Discord website, NOT the Discord client. These encryption techniques sound like voodoo, but it's really not. Discord uses the exact same encryption techniques as Google, your banks, and world governments. 


In regards to point B, the Discord client is largely an Electron application (basically, it's Google's Chromium browser loading a custom webpage) that connects to the Discord website. It does have a lot of extra features that aren't on the Discord website like Windows 10 notifications, but the point is that anyone can already make something fairly similar using Electron just by being a wrapper around the Discord website. From there, all they have to do is bundle it with adware. There's nothing stopping people from doing this now, besides the fact that it's scummy, probably very illegal, and hard to market. 


The thing is, maintaining an open source repository can sometimes be a lot of work, and for a company like Discord that has to make sure their client integrates VERY well with their web server, can't really provide much benefit. Think about it - things like Search run on the Discord website, not on the Discord client. As a result, even if Discord had a million open source developers working on their client, they couldn't integrate Search in the way that it's being done now without a similar team working on the website. 



So what does Discord have to gain by investing in open source project maintenance? I can't really think of much. 


I really hope they do open source it though. That would be cool.

**Comment by sl1pkn07** [comment]
@kessanro
a) Security through obscurity is bad idea. see windows :)
b) if that people is fool and download the programs in 3rd sites... is normal that people will eat that malware
c) would just be a bad idea because of what i talked about :P

**Comment by skwerlman** [comment]
It's closed source because if it were open source, anyone could make discord, and no one would need hammer&chisel. they need it to be proprietary in order to make a profit while still providing the cool things that make discord so awesome, like instant one-click servers for free. it's literally nothing more than a business decision they had to make because we live in a capitalist society. discord is a product, and giving away the secret to your product is commercial suicide.

i love FOSS, and i'd love to see discord get open-sourced, but until we live in a world where money isn't the main driving force behind advancement (i.e. not capitalism), it will not be commercially viable (with only a few exceptions)

also:
@kessanro you should read [this](https://en.wikipedia.org/wiki/Public-key_cryptography) if you're interested in learning more about how crypto works. Its got a lot of reasons why security can exist even when the attacker knows exactly how the lock (or encryption) works. also, it's got a HUGE list of references at the bottom for further reading.
@sl1pkn07 blaming viruses on the victims is a pretty naive way to look at the world. maybe we should stop patching bugs in software because if people were competent they'd just patch it themselves in raw x86. :P

**Comment by crmarsh** [comment]
I don't think there are any big security risks to releasing the source, but fwiw it is a huge hassle to compile in the current state. I pretty much agree with @gonzalezjo, I just wanted to add that I think the most valuable addition we can make to open source is in contributing to the libraries we use, and open sourcing specific things we make that we think other people might want to use. You can see from [our company's github](https://github.com/hammerandchisel) that is what we are doing now, and I expect we will do more and more over time.

**Comment by Vampire** [comment]
@crmarsh where are those repos now? The org you linked to has no public repos.

**Comment by OvermindDL1** [comment]
> @crmarsh where are those repos now? The org you linked to has no public repos.

Yeah I don't know what is with the link he put, but the discord's repo of various libraries is at (I use a few of them actually, simple but useful): 
 https://github.com/discordapp

**Comment by OvermindDL1** [comment]
> Yeah, the org name changed to discordapp

There should probably be a redirect added to the description so old links at least have a description of where to go instead.  :-)
